executor/runcexecutor/executor_linux.go

```diff
@@ -0,0 +1,163 @@
+		err := w.runc.Exec(ctx, id, *specsProcess, &runc.ExecOpts{
+			PidFile: pidfile,
+			IO:      io,
+		})
+		var exitError *exec.ExitError
+		if errors.As(err, &exitError) {
+			return exitError.ExitCode(), err
+		}
+		if err != nil {
+			return containerd.UnknownExitStatus, err
+		}
+		return 0, nil
+	})
+}
+
+type runcCall func(ctx context.Context, pidfile string, io runc.IO) (int, error)
+
+func (w *runcExecutor) callWithIO(ctx context.Context, id, bundle string, process executor.ProcessInfo, call runcCall) (int, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	pidfile, err := ioutil.TempFile(bundle, "*.pid")
+	if err != nil {
+		return containerd.UnknownExitStatus, errors.Wrap(err, "failed to create pidfile")
+	}
+	defer os.Remove(pidfile.Name())
+	pidfile.Close()
+
+	if !process.Meta.Tty {
+		return call(ctx, pidfile.Name(), &forwardIO{stdin: process.Stdin, stdout: process.Stdout, stderr: process.Stderr})
+	}
+
+	ptm, ptsName, err := console.NewPty()
+	if err != nil {
+		return containerd.UnknownExitStatus, err
+	}
+
+	pts, err := os.OpenFile(ptsName, os.O_RDWR|syscall.O_NOCTTY, 0)
+	if err != nil {
+		ptm.Close()
+		return containerd.UnknownExitStatus, err
+	}
+
+	eg, ctx := errgroup.WithContext(ctx)
+
+	if process.Stdin != nil {
+		eg.Go(func() error {
+			io.Copy(ptm, process.Stdin)
+			return nil
+		})
+	}
+
+	if process.Stdout != nil {
+		eg.Go(func() error {
+			io.Copy(process.Stdout, ptm)
+			return nil
+		})
+	}
+
+	eg.Go(func() error {
+		// need to poll until the pidfile has the pid written to it
+		pidfileCtx, timeout := context.WithTimeout(ctx, 10*time.Second)
+		defer timeout()
+		var execProcess *os.Process
+		for {
+			st, err := os.Stat(pidfile.Name())
+			if err == nil && st.Size() > 0 {
+				pid, err := runc.ReadPidFile(pidfile.Name())
+				if err != nil {
+					return errors.Wrapf(err, "unable to read pid file: %s", pidfile.Name())
+				}
+				execProcess, err = os.FindProcess(pid)
+				if err != nil {
+					return errors.Wrapf(err, "unable to find process for pid %d", pid)
+				}
+				break
+			}
+			select {
+			case <-pidfileCtx.Done():
+				return errors.New("pidfile never updated")
+			case <-time.After(100 * time.Microsecond):
+			}
+		}
+		for {
+			select {
+			case <-ctx.Done():
+				return nil
+			case resize := <-process.Resize:
+				err = ptm.Resize(console.WinSize{
+					Height: uint16(resize.Rows),
+					Width:  uint16(resize.Cols),
+				})
+				if err != nil {
+					logrus.Errorf("failed to resize ptm: %s", err)
+				}
+				err = execProcess.Signal(signal.SIGWINCH)
+				if err != nil {
+					logrus.Errorf("failed to send SIGWINCH to process: %s", err)
+				}
+			}
+		}
+	})
+
+	defer func() {
+		if process.Stdin != nil {
+			process.Stdin.Close()
+		}
+		pts.Close()
+		ptm.Close()
+		cancel() // this will shutdown resize loop
+		err := eg.Wait()
+		if err != nil {
+			logrus.Warningf("error while shutting down tty io: %s", err)
+		}
+	}()
```

Comment:
Prefer `defer` closer to line 78 in case code changes in the future and introduces early exit code paths.

Reply:
updated